de.metas.business/src/main/java/org/adempiere/pricing/api/impl/MDiscountSchemaBL.java

```diff
@@ -77,45 +79,29 @@ public BigDecimal calculateDiscount(final I_M_DiscountSchema schema,
 				bPartnerFlatDiscount);
 	}	// calculateDiscount
 
+
 	@Override
-	public BigDecimal calculateDiscount(final I_M_DiscountSchema schema,
+	public BigDecimal calculateDiscount(@NonNull final I_M_DiscountSchema schema,
 			final BigDecimal qty,
 			final BigDecimal Price,
 			final int M_Product_ID,
 			final int M_Product_Category_ID,
 			final List<I_M_AttributeInstance> instances,
 			final BigDecimal bPartnerFlatDiscount)
 	{
-		final BigDecimal bpFlatDiscountToUse;
-		if (bPartnerFlatDiscount == null)
-		{
-			bpFlatDiscountToUse = Env.ZERO;
-		}
-		else
-		{
-			bpFlatDiscountToUse = bPartnerFlatDiscount;
-		}
-
+		final BigDecimal bpFlatDiscountToUse = bPartnerFlatDiscount == null ? BigDecimal.ZERO : bPartnerFlatDiscount;
 		final String discountType = schema.getDiscountType();
-
 		final boolean isQtyBased = schema.isQuantityBased();
-		//
+
 		if (X_M_DiscountSchema.DISCOUNTTYPE_FlatPercent.equals(discountType))
 		{
-			if (schema.isBPartnerFlatDiscount())
-			{
-				return bpFlatDiscountToUse;
-			}
-
-			return schema.getFlatDiscount();
+			return computeFlatDiscount(schema, bpFlatDiscountToUse);
 		}
-
-		// Not supported
 		else if (X_M_DiscountSchema.DISCOUNTTYPE_Formula.equals(discountType)
 				|| X_M_DiscountSchema.DISCOUNTTYPE_Pricelist.equals(discountType))
 		{
 			logger.info("Not supported (yet) DiscountType=" + discountType);
```

Comment:
* this one looks like a WARN for me... if not an exception, not sure
* also pls don't concatenate strings